text/0000-copy-closures.md

```diff
@@ -0,0 +1,170 @@
+- Feature Name: copy_closures
+- Start Date: 2017-8-27
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Implement `Clone` and `Copy` for closures where possible:
+
+```rust
+// Many closures can now be passed by-value to multiple functions:
+fn call<F: FnOnce()>(f: F) { f() }
+let hello = || println!("Hello, world!");
+call(hello);
+call(hello);
+
+// Many `Iterator` combinators are now `Copy`/`Clone`:
+let x = (1..100).map(|x| x * 5);
+let _ = x.map(|x| x - 3); // moves `x` by `Copy`ing
+let _ = x.chain(y); // moves `x` again
+let _ = x.cycle(); // `.cycle()` is only possible when `Self: Clone`
+
+// Closures which reference data mutably are not `Copy`/`Clone`:
+let mut x = 0;
+let incr_x = || x += 1;
+call(incr_x);
+call(incr_x); // ERROR: `incr_x` moved in the call above.
+
+// `move` closures implement `Clone`/`Copy` if the values they capture
+// implement `Clone`/`Copy`:
+let mut x = 0;
+let print_incr = move || { println!("{}", x); x += 1; };
+
+fn call_three_times<F: FnMut()>(mut f: F) {
+    for i in 0..3 {
+        f();
+    }
+}
+
+call_three_times(print_incr); // prints "0", "1", "2"
+call_three_times(print_incr); // prints "0", "1", "2"
+```
+
+# Motivation
+[motivation]: #motivation
+
+Idiomatic Rust often includes liberal use of closures.
+Many APIs have combinator functions which wrap closures to provide additional
+functionality (e.g. methods in the [`Iterator`] and [`Future`] traits).
+
+However, closures are unique, unnameable types which do not implement `Copy`
+or `Clone`. This makes using closures unergonomic and limits their usability.
+Functions which take closures, `Iterator` or `Future` combinators, or other
+closure-based types by-value are impossible to call multiple times.
+
+One current workaround is to use the coercion from non-capturing closures to
+`fn` pointers, but this introduces unnecessary dynamic dispatch and prevents
+closures from capturing values, even zero-sized ones.
+
+This RFC solves this issue by implementing the `Copy` and `Clone` traits on
+closures where possible.
+
+[`Iterator`]: https://doc.rust-lang.org/std/iter/trait.Iterator.html
+[`Future`]: https://docs.rs/futures/*/futures/future/trait.Future.html
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If a non-`move` closure doesn't mutate captured variables,
+then it is `Copy` and `Clone`:
```

Comment:
It's a nit pick, but -- strictly speaking -- this is not always true. A non-move closure can take ownership of values that it (in turn) moves, and hence might not be copy or clone (if those values are not copy or clone).

Reply:
(Notably, those closures would be `FnOnce`.)